File: 2nd_quarter/week04/whaeun/할인행사.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    map<string, int> shopping_list;
    
    for(int i = 0; i < want.size(); i++){
        shopping_list[want[i]] = number[i];
    }
    
    // 10일 모두 구매 시 answer 증가
    int days = 0;
    
    for(int i = 0; i < discount.size(); i++){
        
        if(days == 10){
            shopping_list[discount[i - 10]]++;
            days--;
        }
        
        // is exist in shopping list?
        if(shopping_list[discount[i]] > 0){
            days++;
            shopping_list[discount[i]]--;
            
            if(days == 10){
                answer++;
            }
        }
        else{
            if(days > 0){
                for(int j = days; j >= 1; j--){
                    shopping_list[discount[i - j]]++;
                    days--;
                    
                    if(shopping_list[discount[i]] > 0){
                        days++;
                        shopping_list[discount[i]]--;
            
                        break;
                    }
                }
            };
        }
    }
    
    return answer;
}

File: 2nd_quarter/week04/rash/할인 행사.py
from collections import Counter
# 풀이 계획
# Counter함수의 사용법을 몰라서 고생했다ㅜ
# want와 number를 딕셔너리화 한 후
# discount를 Counter화 한 것과 값을 비교했다.
# 만약 Counter값이 같다면 answer에 1을 더한다
def solution(want, number, discount):
    answer = 0
    wishList = {}

    for i in range(len(want)):
        wishList[want[i]] = number[i]

    print(wishList)

    for j in range(len(discount) - 9):
        tmp = Counter(discount[j:j + 10])
        if tmp == wishList:
            answer += 1

    return answer

